main/javascript-modifications/lockdown.md

```diff
@@ -0,0 +1,543 @@
+
+### Background
+
+There are two reasons for removing `RegExp.*` methods from usable JavaScript.
+One is for `RegExp.prototype.compile()` and one for all other `RegExp.*()` methods.
+
+- In standard JavaScript, `RegExp.prototype.compile()` may 
+  violate the object invariants of frozen `RegExp` instances. This violates
+  assumptions elsewhere, and so can corrupt other guarantees. For example, 
+  the `Proxy` abstraction preserves the object invariants only if its target 
+  does. If a non-conforming object is available, it can construct an also 
+  non-conforming proxy object.
+
+  The default `'safe'` setting deletes this dangerous method. The
+  `'unsafe'` setting keeps it for compatibility purposes at the price of riskier code.
+
+- In standard JavaScript, legacy `RegExp` static methods like `RegExp.lastMatch()` 
+  are an unsafe global [overt communications channel](https://agoric.com/taxonomy-of-security-issues/).
+  The `RegExp` constructor reveals information derived from the last match
+  made by any `RegExp` instance in a form of non-local causality.
+
+  Except for `RegExp.prototype.compile()` all `RegExp` static methods are currently 
+  removed under **both** the `regExpTaming` `'safe`' and `'unsafe'` settings. This
+  has not caused any compatibility problems; if it does, we may allow a subset 
+  under the `'unsafe'` setting.
+
+## `localeTaming` Options
+
+```js
+lockdown(); // localeTaming defaults to 'safe'
+// or
+lockdown({ localeTaming: 'safe' }); // Alias toLocaleString to toString, etc
+// vs
+lockdown({ localeTaming: 'unsafe' }); // Allow locale-specific behavior
+```
+### Purpose
+
+The default `'safe'` setting replaces each method listed below with their
+corresponding non-locale-specific method. For example, `Object.prototype.toLocaleString()`
+becomes another name for `Object.prototype.toString()`. 
+
+The `'unsafe'` setting keeps the original behavior for compatibility at the price
+of reproducibility and fingerprinting. 
+
+In standard JavaScript, these built-in methods have `Locale` or `locale` in their name
+and are affected by `localeTaming`:
+- `toLocaleString`
+- `toLocaleDateString`
+- `toLocaleTimeString`
+- `toLocaleLowerCase`
+- `toLocaleUpperCase`
+- `localeCompare`
+
+### Background
+
+All of the "locale methods" have, by design, global behavior not fully determined by the
+JavaScript spec. Their behavior varies with location and culture. 
+
+However, by placing this information of shared primordial prototypes, **(tyg todo: I'm not parsing the
+previous phrase. What does "placing this information of shared prototypes" mean, or is there a typo
+or missing text in there?)** 
+it cannot differ per Comparment. So one Compartment cannot virtualize the locale for code
+running in another Compartment. Worse, on some engines the methods' behavior may change at runtime as 
+the machine is "moved" between different locales, i.e., if the operating system's locale is
+reconfigured while JavaScript code is running.
```

Comment:
```suggestion
The behavior of all of these methods varies depending on the user's locale and can change while a program is running. This breaks determinism since a program run in one compartment might not have the same result in another compartment or at another time. Revealing the user's locale also helps unscrupulous programs to track even identify the user without their consent.

The behavior of methods like `toLocaleString` might be acceptable if they could be governed on a per-compartment basis ("virtualized"), but because these methods appear on the prototype of shared intrinsic objects like the String prototype, there is no safe alternative design.

Instead, the hosting program can choose to reveal the locale to all compartments by setting `localeTaming` to `unsafe`, or inject a `locale` object into selected compartments with the powers of these methods.
```